src/ucp/core/ucp_am.c

```diff
@@ -1330,10 +1332,10 @@ ucp_am_handle_unfinished(ucp_worker_h worker, ucp_recv_desc_t *first_rdesc,
      *                                                       needed anymore,
      *                                                       can overwrite)
      */
-    msg                = UCS_PTR_BYTE_OFFSET(first_rdesc + 1,
-                                             first_rdesc->payload_offset);
-    first_rdesc        = (ucp_recv_desc_t*)msg - 1;
-    first_rdesc->flags = UCP_RECV_DESC_FLAG_MALLOC;
+    desc_offset                 = first_rdesc->payload_offset;
+    first_rdesc                 = (ucp_recv_desc_t*)msg - 1;
+    first_rdesc->flags          = UCP_RECV_DESC_FLAG_MALLOC;
+    first_rdesc->payload_offset = desc_offset;
```

Comment:
why isn't it 0?
```
new desc layout:                         |desc|data|
```
this means data is right after desc,
and:
```
    uint32_t                payload_offset;  /* Offset from end of the descriptor
                                              * to AM data */
```

Reply:
i reuse `payload_offset` field to be able to free originally allocated rdesc

Reply:
so the new usage is not in sync with documentation?

Reply:
in fact it is in sync. I keep here offset from the rdesc to the payload. And later, during release phase I step back to get correct pointer to the original rdec

Reply:
So later, during release phase, "payload_offset" has different meaning?
should use union then

Reply:
no, it has the same meaning - the distance between rdesc and payload

Reply:
according to the comment
`new desc layout:                         |desc|data|`
the new distance is 0
so which one is correct?

Reply:
this is the new desc, which we have to introduce instead of old one, because there are headers between the old descriptor and payload. And we copy some of the old fields to this new rdesc, to correctly release it: flags and payload_offset.
if we create a union, how would it be  more clear? What would prevent the user to use incorrect field?

Reply:
There should be consistency between field name, documentation, and its usage.
If we need to know the distance from rdesc to beginning of mpool object/malloc'd buffer, need to add new field, probably union with uct_desc_offset. for example "ucp_mdesc_offset", and document its purpose and usage.
Then, we can prevent using incorrect field, because the intended usage will be clear during code development and review.

Reply:
imo, this adds additional complexity rather than consistency, but ok will create union